Add THREE.CompressedArrayTexture

RenaudRohlinger opened the pull request:
WebGLTextures part should work, I'm just very confused with the KTXLoader part and on how to create a proper dataset with toktx.

Right now I am using the following command `toktx --layers 2 --t2 textures.ktx2 hair_m_baseColor.png hair_m_normal.png` in the `examples/models/gltf/sougen/` folder.

I suppose that the support of this should work:
https://developer.mozilla.org/en-US/docs/Web/API/WebGL_API/Compressed_texture_formats

https://github.com/KhronosGroup/KTX-Software/pull/468/files

This might interest 😄 @donmccurdy

RenaudRohlinger committed:
wip

RenaudRohlinger committed:
Merge branch 'mrdoob:dev' into texture2darray_compressed

donmccurdy reviewed (commented):
Thanks @RenaudRohlinger, great to see this! 👍

I don't think I know enough to comment on the renderer internals, perhaps @Mugen87 can say more.

> Right now I am using the following command toktx --layers 2 --t2 textures.ktx2 hair_m_baseColor.png hair_m_normal.png in the examples/models/gltf/sougen/ folder.

Careful on this – these texture types have different encodings, so one or the other is likely to be incorrectly-encoded or to have lost precision. Probably want to keep color and non-color textures in separate arrays, and use `--assign_oetf` accordingly.

RenaudRohlinger committed:
texture array is working with hardcoded values

RenaudRohlinger commented on examples/jsm/loaders/KTX2Loader.js:
```diff
@@ -622,6 +640,7 @@ const FORMAT_MAP = {
 	[ VK_FORMAT_R32G32B32A32_SFLOAT ]: RGBAFormat,
 	[ VK_FORMAT_R16G16B16A16_SFLOAT ]: RGBAFormat,
 	[ VK_FORMAT_R8G8B8A8_UNORM ]: RGBAFormat,
+	[ VK_FORMAT_R8G8B8_SRGB ]: RGBAFormat,
```
This was generated from a specific type of export and I don't really remember the exact instructions I will need to investigate more deeply but yes I agree let's not go back to RGB 😄

RenaudRohlinger commented on examples/jsm/loaders/KTX2Loader.js:
```diff
@@ -733,8 +751,7 @@ async function createDataTexture( container ) {
 	//
-
-	const texture = pixelDepth === 0
+	const texture = layerCount > 1 ? new DataArrayTexture( view, pixelWidth, pixelHeight, layerCount ) : pixelDepth === 0
```
I did test both but I think it would be better to focus only on CompressedTexture first 👍

RenaudRohlinger commented:
Good news @donmccurdy! I made some progress on this feature and I can make the example work with some hard-coded data. In the example, the right part of the hair reflects layer 0, and the left part is layer 1.
 
I managed to isolate the issues and I think the remaining part is inside the ktx transcoded. The array buffer data of the transcoded image will only output the first layer.

I made a very basic compressed array texture of 4x4 px red on layer 0 and 4x4 px blue on layer 1 to make these tests easier.

There is a bunch of comments inside the KTXParser file with my catch so far but my knowledge is very limited in this area so help would be more than welcome 😄 


<img width="1723" alt="image" src="https://user-images.githubusercontent.com/15867665/193977288-f6273cec-6d6c-46db-9c5e-246f279ebafe.png">

donmccurdy commented:
Related:

- https://github.com/mrdoob/three.js/issues/19676 (/cc @gkjohnson)

The Basis transcoder API requires `layerIndex` arguments for the [`transcodeImage`](https://github.com/BinomialLLC/basis_universal/blob/249f09d5f2c312b03fe6e0869fef343e18641cff/webgl/transcoder/basis_wrappers.cpp#L1499-L1502) and [`getImageLevelInfo`](https://github.com/BinomialLLC/basis_universal/blob/249f09d5f2c312b03fe6e0869fef343e18641cff/webgl/transcoder/basis_wrappers.cpp#L1655) methods, which we will need. Currently those are unused by the implementation. Once we have data for each layer, I am not sure what to do with it... do we just concatenate the data for all layers at the same mipmap level? I haven't found any examples yet.

I think we are going to need a THREE.CompressedArrayTexture class for this implementation. Adding THREE.Compressed3DTexture might also be nice but is not required here. The main difference is in the mipmap layouts.

donmccurdy committed:
KTX2Loader: Decode additional layers from array textures.

donmccurdy commented:
> do we just concatenate the data for all layers at the same mipmap level?

I guess that's right — pushed https://github.com/mrdoob/three.js/pull/24745/commits/32c1a67e2de98d54323975ce0fa63ce69b5998b3.

![Screen Shot 2022-10-05 at 11 34 48 AM](https://user-images.githubusercontent.com/1848368/194101753-f6f88767-0292-4a6f-9011-b5443d414e96.png)

RenaudRohlinger committed:
added compressedarraytexture

RenaudRohlinger committed:
Merge remote-tracking branch 'upstream/dev' into texture2darray_compressed

RenaudRohlinger commented:
Added CompressedArrayTexture, updated the example, and cleaned up some code.

Should we try to support Compressed3DTexture at the same time?

Also maybe it's worth mentioning that compressed array texture generates a bunch of warnings on firefox that should be ignored.
https://stackoverflow.com/questions/57734645/error-tex-image-texture-2d-level-0-is-incurring-lazy-initialization-on-webgl/57734917#57734917

>  So in other words, getting Firefox to shut up about its warning actually makes your code much slower and use much more memory then if firefox just silently cleared the texture on its own.

![image](https://user-images.githubusercontent.com/15867665/194284127-acf7571f-ea7c-492d-8729-a52fb7b08fe8.png)

RenaudRohlinger committed:
fix merge

RenaudRohlinger committed:
fixed level issues

donmccurdy commented on src/textures/CompressedArrayTexture.js:
```diff
@@ -0,0 +1,18 @@
+
+class CompressedArrayTexture extends CompressedTexture {
+
+	constructor( depth, mipmaps, width, height, format, type, mapping, wrapS, wrapT, magFilter, minFilter, anisotropy, encoding ) {
```
Let's follow the order of the ArrayTexture constructor here...

https://github.com/mrdoob/three.js/blob/d081c5a3501d272d19375fab1b01fedf9df29b22/src/textures/DataArrayTexture.js#L6

I would suggest skipping the arguments after `depth` (layers) for now. It is easy to add them later, but hard to change the order once they're added, and we may need to think about the order a bit for compatibility with other texture classes.

```javascript
constructor( mipmaps, width, height, depth ) { ... }
```

donmccurdy commented on src/textures/CompressedArrayTexture.js:
```diff
@@ -0,0 +1,18 @@
+
+		super( mipmaps, width, height, format, type, mapping, wrapS, wrapT, magFilter, minFilter, anisotropy, encoding );
+
+		this.isDataArrayTexture = true;
```
Can we make this work with `this.isCompressedArrayTexture = true;` instead? I think it will cause confusion later if we misrepresent the type here.

donmccurdy reviewed (commented):
> Should we try to support Compressed3DTexture at the same time?

Let's consider this in a future PR. Updating KTX2Loader, WebGLRenderer, adding CompressedArrayTexture, and adding an example is already a fair bit for one PR!

The changes in KTX2Loader to support compressed array textures look good to me. I am not sure I can review the changes to WebGLState and WebGLTextures, @Mugen87 do you mind reviewing this part?

RenaudRohlinger committed:
fix copyTextureToTexture3D, updated compressedarraytexture arguments, and added doc

RenaudRohlinger committed:
remove build

RenaudRohlinger commented:
Thanks for all the feedback!

I updated CompressedArrayTexture, the code in WebGLTextures accordingly, updated copyTextureToTexture3D in the renderer too, and added the documentation.

Should be ready for review 👍

RenaudRohlinger committed:
fix deepscan

CodyJasonBennett commented on examples/jsm/loaders/KTX2Loader.js:
```diff
@@ -236,16 +240,30 @@ class KTX2Loader extends Loader {
+			texture.isDataArrayTexture = true;
+			texture.wrapR = ClampToEdgeWrapping;
+			texture.image.depth = layerCount;
+			texture.unpackAlignment = 1;
```
https://webglfundamentals.org/webgl/lessons/webgl-data-textures.html would explain this better after the first example. I hardcode this to 1 in my WebGL projects to play nice with data textures.

donmccurdy replied:
Doesn't seem like we should be setting alignment-related properties for a _compressed_ texture though, or if so, the padding should be specified somewhere in the file?

In any case it looks like this has been removed in the latest draft 👍

donmccurdy commented on src/renderers/webgl/WebGLTextures.js:
```diff
@@ -671,7 +671,7 @@ function WebGLTextures( _gl, extensions, state, properties, capabilities, utils,
 
```
Further down in this file, do we need to add `wrapR` here?

```javascript
	function getTextureCacheKey( texture ) {

		const array = [];

		array.push( texture.wrapS );
		array.push( texture.wrapT );
		array.push( texture.magFilter );
		array.push( texture.minFilter );
		// ...
```

Mugen87 replied:
Good catch! I guess `array.push( texture.wrapR || 0 );` should do it.

Mugen87 replied:
Updated!

Mugen87 committed:
Update WebGLTextures.js

Mugen87 commented on examples/webgl2_texture2darray_compressed.html:
```diff
@@ -0,0 +1,313 @@
+
+				function obc( material ) {
+
+					material.onBeforeCompile = function ( shader ) {
```
Sorry for the late feedback but I have encountered this section just now.

Would it be possible to showcase compressed array textures differently and not by patching built-in materials? In the past, we made bad experiences with such examples since they are hard to maintain (they easily break when we refactor shaders). Could you use a custom shader material with minimal code (similar to https://threejs.org/examples/webgl2_buffergeometry_attributes_integer)?

donmccurdy replied:
Simplest option here might be to copy https://threejs.org/examples/?q=array#webgl2_materials_texture2darray using THREE.CompressedArrayTexture? That example probably makes it more clear how the array texture is being used. 

Not opposed to doing something different here, but also don't want to hold the PR back.

Mugen87 committed:
Update webgl2_texture2darray_compressed.html

Mugen87 committed:
Update webgl2_texture2darray_compressed.html

RenaudRohlinger committed:
replaced with simple example

RenaudRohlinger committed:
remove build

RenaudRohlinger commented:
I replaced the example with something a lot more basic. I'm using 6 frames from a Ghibli movie, I have absolutely no idea what's the boundaries about licenses here, I suppose it could be problematic? In that case, any cc0 loop would be welcome as I can't really find anything looking decent enough for the example section.

![image](https://user-images.githubusercontent.com/15867665/196071511-e0298e33-c17c-4841-b182-1805f5b8536b.png)

Mugen87 commented:
> I suppose it could be problematic?

I'm not a lawyer but I guess it's the same like when the clip would appear in a movie review video or something similar. It should fall under the *fair use* area.

Mugen87 committed:
Update webgl2_texture2darray_compressed.html

Mugen87 committed:
Update webgl2_texture2darray_compressed.html

RenaudRohlinger committed:
make the example framerate independent

RenaudRohlinger committed:
fix init texture2darray for compressed texture in webglrenderer

Mugen87 committed:
Update webgl2_texture2darray_compressed.html
